### ISSUE TITLE
Expose FileReader & way to create File instances

### ISSUE THREAD

**Issue by Gozala** [RangerMauve/webrun#15]
At the moment there is no way to read local files without resorting to node's require. I think it would be awesome if there was a way to use [`FileReader`](https://developer.mozilla.org/en-US/docs/Web/API/FileReader) and [`File`](https://developer.mozilla.org/en-US/docs/Web/API/File) APIs for that purpose. Although one would obviously need a way to create a `File` instance from the file URL somehow.

It might be worth consulting https://wicg.github.io/writable-files/ spec.

I would also really love if we could somehow do a better job on sandboxing. My immediate idea would be to have another argument for webrun that would restrict fs access to specific paths. For example:

```
webrun main.js --read+write ./store/
```

That would allow `main.js` to read write under the given path but nowhere else. It could further be restricted to the code pulled in for example:

```js
import staticServer from "./server?read=./store"
```

`staticServer` would not be capable of writing files anywhere but will be able to read from `./store`.

It might be also worth checking how https://deno.land/ is going about it.

**Comment by RangerMauve** [comment]
What do you think about a manifest or something for permissions?

**Comment by RangerMauve** [comment]
I really want to add the new files API, but I don't think there's a node.js implementation available yet.

**Comment by Gozala** [comment]
> What do you think about a manifest or something for permissions?

Manifest is usually per application, but what if you want to use some module without giving it access to your filesystem ?

**Comment by Gozala** [comment]
> I really want to add the new files API, but I don't think there's a node.js implementation available yet.

I hacked something together for my use case for now

**Comment by RangerMauve** [comment]
What about something like `--allow-read` and `--allow-write` which default to reading from the local directory or setting `--allow-read=./store` for a specific directory?

**Comment by Gozala** [comment]
> What about something like `--allow-read` and `--allow-write` which default to reading from the local directory or setting `--allow-read=./store` for a specific directory?

I'm honestly not sure. It's also unsure if sandboxing per module even makes sense, usually it's per process so maybe that's a better way to go about it.

I also looked at deno & approach there is pretty interesting. If you attempt to start a server it will request permission in CLI & you can provide CLI arg to grant permissions ahead of time, see:
https://github.com/denoland/deno/blob/master/Docs.md#tcp-echo-server

That is probably a good model and matches more or less the web.

**Comment by RangerMauve** [comment]
Yeah, I was thinking per process, not module. Prompting is a pretty good idea. With regards to directories, I could see it being pretty annoying if it's asked a lot. Maybe you need to specify a path in the writable files API and that's when it asks?

**Comment by Gozala** [comment]
Here is the list of APIs that I would really like to see that are related to working with files

- [ ] [`Blob`](https://developer.mozilla.org/en-US/docs/Web/API/Blob)
- [ ] [`File`](https://developer.mozilla.org/en-US/docs/Web/API/File) API allows file construction & reading. It also can be user as response / request body.

  > I think one way to possibly expose FS in web compatible manner would be through something like `self.postMessage({ type:"fs.write", url: new URL("file:///...."), file: new File(...) })`.

- [ ] [`FileReader`](https://developer.mozilla.org/en-US/docs/Web/API/FileReader) Not a great API but at least web compatible. It also seems to be embraced by https://wicg.github.io/writable-files/
- [] [`FileReaderSync`](https://developer.mozilla.org/en-US/docs/Web/API/FileReaderSync) Not sure about this one, but node allow sync IO and browsers do allow it in web workers, so it might be reasonable request.

**Comment by RangerMauve** [comment]
I don't think the postmessage stuff is necessary, I think that the writable-files API scoped to the `pwd` would be enough for most applications.

**Comment by RangerMauve** [comment]
Do you know if any part of the `writable files` or `File` APIs work with the new WHATWG streams?
